src/Avalonia.Controls/VirtualizingStackPanel.cs

```diff
@@ -834,7 +834,7 @@ public IReadOnlyList<double> GetIrregularSnapPoints(Orientation orientation, Sna
                     {
                         var averageElementSize = EstimateElementSizeU();
                         double snapPoint = 0;
-                        for (var i = 0; i < Items.Count; i++)
+                        for (var i = 0; i < itemCount; i++)
                         {
                             var container = ContainerFromIndex(i);
```

Comment:
`ContainerFromIndex` will access the items source too as it needs to check if the index represents an item that is its own container. We can go directly to `_realizedElements` or use `GetRealizedElement(int)` to skip that check.